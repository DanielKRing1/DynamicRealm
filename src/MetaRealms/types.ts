import { OptionalExceptFor } from '../types/types';

export type LoadableRealmParams = {
    metaRealmPath: string;
    loadableRealmPath: string;
};
export type LoadableSchemaParams = {
    metaRealmPath: string;
    schemaName: string;
};
export type SaveSchemaParams = {
    schema: Realm.ObjectSchema;
} & LoadableRealmParams;
export type UpdateSchemaParams = {
    schema: Realm.ObjectSchema;
} & LoadableRealmParams;
export type RmSchemaParams = {
    schemaName: string;
} & LoadableRealmParams;
export type LoadRealmParams = {
    schemaNames: string[];
} & LoadableRealmParams;

export type LoadableSchemaRowProperties = {
    name: string;
    schema: string;
    metadata: string;
};
export type LoadableSchemaRow = LoadableSchemaRowProperties & Realm.Object;

export type LoadableRealmRowProperties = {
    realmPath: string;
    schemaNames: string[];
    schemaVersion: number;
};
export type LoadableRealmRow = LoadableRealmRowProperties & Realm.Object;
export type OptionalLoadableRealmRowProperties = OptionalExceptFor<LoadableRealmRowProperties, 'realmPath'>;
