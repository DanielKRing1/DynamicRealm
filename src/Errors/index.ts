export const INIT_ERROR = new Error('Please init DynamicRealm (with dynamicRealm.init()) before using it!');
export const METADATA_ARRAY_INTEGER_ERROR = new Error('Please provide an integer value to access your metadata');
