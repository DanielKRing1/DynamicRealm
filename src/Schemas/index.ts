export const DYNAMIC_SCHEMA_NAME: string = 'Dynamic';
export const DynamicSchema: Realm.ObjectSchema = {
    name: DYNAMIC_SCHEMA_NAME,
    primaryKey: 'name',
    properties: {
        name: 'string',
        primaryKey: 'string',
        realmName: 'string',
        schema: 'string',
        metadata: 'string',
    },
};

export const DYNAMIC_REALM_NAME: string = 'Realms';
export const DynamicRealm: Realm.ObjectSchema = {
    name: DYNAMIC_REALM_NAME,
    primaryKey: 'name',
    properties: {
        name: 'string',
        realmPath: 'string',
        schemaNames: 'string[]',
        schemaVersion: { type: 'int', default: 0 },
    },
};
