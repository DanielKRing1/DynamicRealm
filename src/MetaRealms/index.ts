import { Results } from 'realm';
import { CREATE_LOADABLE_REALM_SCHEMA, LoadableRealm, LOADABLE_REALM_TABLE_NAME, LOADABLE_SCHEMA_TABLE_NAME } from '../StaticSchemas';
import MetaRealmManager from './metaRealmManager';
import { LoadableRealmParams, LoadableRealmRow, LoadableRealmRowProperties, LoadableSchemaParams, LoadableSchemaRow, LoadableSchemaRowProperties, RmSchemaParams, SaveSchemaParams, UpdateSchemaParams } from './types';

/**
 * Create new LoadableSchema row and
 * Add schemaName to LoadableRealm row's list of schemaNames if not already included
 * 
 * @param param0 
 */
export const saveSchema = ({ metaRealmPath, loadableRealmPath, schema }: SaveSchemaParams): void => {
    // 1. Get MetaRealm
    const metaRealm: Realm = MetaRealmManager.getMetaRealm(metaRealmPath);

    // 2. Create MetaSchema object to save
    const schemaObj: LoadableSchemaRowProperties = {
        // 2.1. Record the 'name' property for simple querying
        name: schema.name,
        // 2.2. Stringify the schema object
        schema: JSON.stringify(schema),
        // 2.3. Init the metadata as empty
        metadata: '{}',
    };

    // 4. Add the new schema to the MetaSchema table
    _write(metaRealm, () => {
        metaRealm.create(LOADABLE_SCHEMA_TABLE_NAME, schemaObj);
    })

    // 5. Get or create the MetaRealm
    const loadableRealmRow: LoadableRealmRow = getLoadableRealmRow({ metaRealmPath, loadableRealmPath });

    // 6. Add new MetaSchema's name to MetaRealm
    // Do not re-push schema name into LoadableRealm row's list of schemaNames
    _write(metaRealm, () => {
        // @ts-ignore
        // Realm stores values as Set but returns them as list
        if(loadableRealmRow !== undefined && !loadableRealmRow.schemaNames.has(schema.name)) loadableRealmRow.schemaNames.add(schema.name);
    })
}

export const updateSchema = ({ metaRealmPath, loadableRealmPath, schema }: UpdateSchemaParams): boolean => {
    // 1. Get LoadableRealmRow
    const loadableSchemaRow: LoadableSchemaRowProperties = getLoadableSchemaRow({ metaRealmPath, schemaName: schema.name });
    // 2. Get LoadableRealmRow
    const loadableRealmRow: LoadableRealmRowProperties = getLoadableRealmRow({ metaRealmPath, loadableRealmPath });

    const metaRealm: Realm = MetaRealmManager.getMetaRealm(metaRealmPath);

    _write(metaRealm, () => {
        // 3. Update schema
        if(loadableSchemaRow !== undefined) loadableSchemaRow.schema = JSON.stringify(schema);

        // 4. Increment schemaVersion
        if(loadableRealmRow !== undefined) loadableRealmRow.schemaVersion += 1;
    });
    
    return loadableSchemaRow !== undefined && loadableRealmRow !== undefined;
}

/**
 * Delete LoadableSchema row and
 * Remove schema names from LoadableREalm row's list of schemaNames
 * 
 * @param param0 
 * @returns 
 */
export const rmSchema = ({ metaRealmPath, loadableRealmPath, schemaName }: RmSchemaParams): boolean => {
    // 1. Get LoadableRealmRow
    const loadableSchemaRow: LoadableSchemaRowProperties = getLoadableSchemaRow({ metaRealmPath, schemaName });
    // 2. Get LoadableRealmRow
    const loadableRealmRow: LoadableRealmRowProperties = getLoadableRealmRow({ metaRealmPath, loadableRealmPath });

    const metaRealm: Realm = MetaRealmManager.getMetaRealm(metaRealmPath);

    _write(metaRealm, () => {
        // 3. Delete LoadableSchema row
        if(loadableSchemaRow !== undefined) metaRealm.delete(loadableSchemaRow);

        if(loadableRealmRow !== undefined) {
            // @ts-ignore
            // Realm stores values as Set but returns them as list
            // 4. Remove LoadableSchema name from LoadableRealm row's schemaNames
            loadableRealmRow.schemaNames.delete(schemaName);

            // 5. Increment schemaVersion
            loadableRealmRow.schemaVersion += 1;
        }
    });

    return loadableSchemaRow !== undefined && loadableSchemaRow !== undefined;
}

export const getAllLoadableSchemaRows = (metaRealmPath: string): Realm.Results<LoadableSchemaRow> => {
    // 1. Get MetaRealm
    const metaRealm: Realm = MetaRealmManager.getMetaRealm(metaRealmPath);
    
    // 2. Get LoadableSchema rows
    const loadableSchemaRows: Results<LoadableSchemaRow> = metaRealm.objects(LOADABLE_SCHEMA_TABLE_NAME);

    return loadableSchemaRows;
}

/**
 * Get LoadableSchema row

* LoadableSchema Table
 * 0..* row of < PK: schemaName, schema data... >
 * 
 * @param param0 
 * @returns 
 */
export const getLoadableSchemaRow = ({ metaRealmPath, schemaName }: LoadableSchemaParams): LoadableSchemaRow | undefined => {
    // 1. Get MetaRealm
    const metaRealm: Realm = MetaRealmManager.getMetaRealm(metaRealmPath);
    
    // 2. Get LoadableSchema row
    const loadableSchemaRow: LoadableSchemaRow | undefined = metaRealm.objectForPrimaryKey(LOADABLE_SCHEMA_TABLE_NAME, schemaName);

    return loadableSchemaRow;
}

export const getAllLoadableRealmRows = (metaRealmPath: string): Realm.Results<LoadableRealmRow> => {
    // 1. Get MetaRealm
    const metaRealm: Realm = MetaRealmManager.getMetaRealm(metaRealmPath);

    // 2. Get LoadableRealm rows
    const loadableRealmRows: Results<LoadableRealmRow> = metaRealm.objects<LoadableRealmRow>(LOADABLE_REALM_TABLE_NAME);

    return loadableRealmRows;
}


/**
 * This method will create the default LoadableRealm row when executed
 * 
 * Get LoadableRealm row and create it with loadableRealmPath PK if not exists
 * 
 * LoadableRealm Table
 * 0..* rows of < PK: loadableRealmPath, schemasToLoad: [schemaNames], schemaVersion: number >
 * 
 * @param param0 
 * @returns 
 */
export const getLoadableRealmRow = ({ metaRealmPath, loadableRealmPath }: LoadableRealmParams): LoadableRealmRow => {
    // 1. Get MetaRealm
    const metaRealm: Realm = MetaRealmManager.getMetaRealm(metaRealmPath);
    
    // 2. Get LoadableRealm row
    let loadableRealmRow: LoadableRealmRow | undefined = metaRealm.objectForPrimaryKey(LOADABLE_REALM_TABLE_NAME, loadableRealmPath);

    // 3. Create LoadableRealm row if not exists
    if (!loadableRealmRow) {
        // 3.1. Create object
        const loadableRealmSchemaObj: LoadableRealmRowProperties = CREATE_LOADABLE_REALM_SCHEMA({ realmPath: loadableRealmPath });

        // 3.2. Save
        _write(metaRealm, () => {
            loadableRealmRow = metaRealm.create(LOADABLE_REALM_TABLE_NAME, loadableRealmSchemaObj);
        });
    }

    return loadableRealmRow as LoadableRealmRow;
}

export const _write = (metaRealm: Realm, writeCb: () => void) => {
    try {
        metaRealm.write(writeCb);
    }
    catch(err) {
        console.log(err)
        // Error thrown to prevent writing duplicate; Do nothing
    }
}
