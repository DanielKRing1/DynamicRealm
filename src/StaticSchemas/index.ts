import { OptionalLoadableRealmRowProperties } from "../MetaRealms/types";

/**
 * SCHEMAS THAT ARE SAVED AND THAT CAN BE LOADED
 */
export const LOADABLE_SCHEMA_TABLE_NAME: string = 'LOADABLE_SCHEMAS';
export const LoadableSchema: Realm.ObjectSchema = {
    name: LOADABLE_SCHEMA_TABLE_NAME,
    primaryKey: 'name',
    properties: {
        name: 'string',
        schema: 'string',
        metadata: 'string',
    },
};

/**
 * GROUPINGS FOR LOADABLE SCHEMAS:
 * WHAT SET OF SCHEMAS WILL BE LOADED INTO A REALM
 */
export const LOADABLE_REALM_TABLE_NAME: string = 'LOADABLE_REALMS';
export const LoadableRealm: Realm.ObjectSchema = {
    name: LOADABLE_REALM_TABLE_NAME,
    primaryKey: 'realmPath',
    properties: {
        realmPath: 'string',
        schemaNames: { type: 'string<>', default: [] },
        schemaVersion: { type: 'int', default: 0 },
    },
};

// @ts-ignore
// Realm expects an array to initialize the set
export const CREATE_LOADABLE_REALM_SCHEMA = ({ realmPath, schemaNames = [], schemaVersion = 0 }: OptionalLoadableRealmRowProperties) => {
    return {
        realmPath,
        schemaNames,
        schemaVersion,
    };
};
