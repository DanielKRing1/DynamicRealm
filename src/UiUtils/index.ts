import { getAllLoadableRealmRows, getLoadableRealmRow, getLoadableSchemaRow } from '../MetaRealms';
import { LoadableRealmRow, LoadableSchemaRow } from '../MetaRealms/types';

export const getLoadableRealmPaths = async (metaRealmPath: string): Promise<string[]> => {
    const allLoadableRealmRows: Realm.Results<LoadableRealmRow> = await getAllLoadableRealmRows(metaRealmPath);
    return allLoadableRealmRows.map((realm) => realm.realmPath);
}

export const getSchemaNames = async (metaRealmPath: string, loadableRealmPath: string): Promise<string[]> => {
    // 1. Get LoadableRealm
    const loadableRealmRow: LoadableRealmRow = await getLoadableRealmRow({ metaRealmPath, loadableRealmPath });

    return loadableRealmRow !== undefined ? Array.from(loadableRealmRow.schemaNames) : [];
}

export const getProperties = async (metaRealmPath: string, schemaName: string): Promise<Realm.PropertiesTypes> => {
    const loadableSchemaRow: LoadableSchemaRow = await getLoadableSchemaRow({ metaRealmPath, schemaName });

    return loadableSchemaRow !== undefined ? JSON.parse(loadableSchemaRow.schema).properties : {};
}