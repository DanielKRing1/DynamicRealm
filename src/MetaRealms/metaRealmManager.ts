import Realm from 'realm';

import { Dict } from "..";
import { LoadableRealm, LoadableSchema } from '../StaticSchemas';

function createMetaRealmsManager() {
    const metaRealmsMap: Dict<Realm> = {};

    /**
     * Get a MetaRealm or
     * Open a new one if one does not exist or it is closed
     * 
     * @param metaRealmPath 
     * @returns 
     */
    const getMetaRealm = async (metaRealmPath: string): Promise<Realm> => {
        // 1. Open Realm if not exists or is closed
        if(!hasOpenMetaRealm(metaRealmPath)) {
            let _realmInstance: Realm = await Realm.open({ schema: [ LoadableRealm, LoadableSchema ], path: metaRealmPath });
            metaRealmsMap[metaRealmPath] = _realmInstance;
        }

        // 2. If still not exists, throw an error
        if(!hasOpenMetaRealm(metaRealmPath)) _throwNoRealmError(metaRealmPath);

        // 3. Return open Realm
        return metaRealmsMap[metaRealmPath];
    };

    /**
     * Close open MetaRealm
     * Then delete from MetaRealm map
     * 
     * @param metaRealmPath 
     */
    const closeMetaRealm = (metaRealmPath: string) => {
        if(hasOpenMetaRealm(metaRealmPath)) {
            const realm: Realm = metaRealmsMap[metaRealmPath];
            realm.close();
        }

        if(hasMetaRealm(metaRealmPath)) delete metaRealmsMap[metaRealmPath];
    }

    const closeAll = () => {
        const metaRealmPaths: string[] = getAllMetaRealmPaths();
        for(let metaRealmPath of metaRealmPaths) {
            closeMetaRealm(metaRealmPath);
        }
    }

    const hasMetaRealm = (metaRealmPath: string): boolean => !!metaRealmsMap[metaRealmPath];

    const hasOpenMetaRealm = (metaRealmPath: string): boolean => {
        if(hasMetaRealm(metaRealmPath)) {
            const realm: Realm = metaRealmsMap[metaRealmPath];
            return !realm.isClosed;
        }

        return false;
    }

    const getAllMetaRealmPaths = () => Object.keys(metaRealmsMap);
    const getAllMetaRealms = () => Object.values(metaRealmsMap);

    const getAllOpenMetaRealmPaths = () => getAllMetaRealmPaths().filter((metaRealmName) => !metaRealmsMap[metaRealmName].isClosed);
    const getAllOpenMetaRealms = () => getAllMetaRealms().filter((metaRealm: Realm) => !metaRealm.isClosed);

    // INTERNAL UTILS

    const _throwNoRealmError = (metaRealmPath: string) => {
        throw new Error(`A Realm does not exist at "${metaRealmPath}", and one could not be opened`);
    }

    return {
        getMetaRealm,
        hasMetaRealm,
        hasOpenMetaRealm,
        closeMetaRealm,
        closeAll,

        getAllMetaRealmPaths,
        getAllMetaRealms,
        getAllOpenMetaRealmPaths,
        getAllOpenMetaRealms,
    };
};

export default createMetaRealmsManager();
