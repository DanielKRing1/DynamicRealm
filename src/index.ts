import { saveSchema, rmSchema, updateSchema } from './MetaRealms';
import LoadableRealmManager from './LoadableRealms/loadableRealmManager';
import { getLoadableRealmPaths, getProperties, getSchemaNames } from './UiUtils';

export * from './MetaRealms/types';
export * from './types/types';

export default {
    saveSchema,
    updateSchema,
    rmSchema,

    ...LoadableRealmManager,

    getLoadableRealmPaths,
    getSchemaNames,
    getProperties,
};
