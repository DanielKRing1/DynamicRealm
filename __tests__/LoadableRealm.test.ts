import fs from 'fs';
import MetaRealmManager from '../src/MetaRealms/metaRealmManager';
import LoadableRealmManager from '../src/LoadableRealms/LoadableRealmManager';
import { saveSchema } from '../src/MetaRealms';

const TEST_NAME: string = 'LoadableRealm';
const TEST_DIRECTORY: string = `__tests__/${TEST_NAME}`;
const META_REALM_PATH1: string = `${TEST_DIRECTORY}/MetaRealm1.path`;
const META_REALM_PATH2: string = `${TEST_DIRECTORY}/MetaRealm2.path`;
const LOADABLE_REALM_PATH1: string = `LoadableRealm1.path`;
const LOADABLE_REALM_PATH2: string = `LoadableRealm2.path`;

const LOADABLE_SCHEMA_1: Realm.ObjectSchema = {
    name: 'LoadableSchema1',
    primaryKey: 'name',
    properties: {
        name: 'string',
        schema1Data: 'string',
    },
};

const LOADABLE_SCHEMA_2: Realm.ObjectSchema = {
    name: 'LoadableSchema2',
    primaryKey: 'name',
    properties: {
        name: 'string',
        schema2Data: 'string',
    },
};

const LOADABLE_SCHEMA_3: Realm.ObjectSchema = {
    name: 'LoadableSchema3',
    primaryKey: 'name',
    properties: {
        name: 'string',
        schema3Data: 'string',
    },
};

const LOADABLE_SCHEMA_4: Realm.ObjectSchema = {
    name: 'LoadableSchema4',
    primaryKey: 'name',
    properties: {
        name: 'string',
        schema4Data: 'string',
    },
};

const LOADABLE_SCHEMA_5: Realm.ObjectSchema = {
    name: 'LoadableSchema5',
    primaryKey: 'name',
    properties: {
        name: 'string',
        schema5Data: 'string',
    },
};

describe('LoadableRealmManager', () => {
    beforeAll(() => {
        if (fs.existsSync(TEST_DIRECTORY)) fs.rmSync(TEST_DIRECTORY, { recursive: true });

        fs.mkdirSync(TEST_DIRECTORY);
    });

    it('Should fail to open a non-existant LoadableRealm', async () => {
            await expect(LoadableRealmManager.loadRealm({ metaRealmPath: META_REALM_PATH1, loadableRealmPath: LOADABLE_REALM_PATH1 })).rejects.toThrowError();
            await expect(LoadableRealmManager.loadRealm({ metaRealmPath: META_REALM_PATH1, loadableRealmPath: LOADABLE_REALM_PATH1 })).rejects.toThrowError();

            await expect(LoadableRealmManager.reloadRealm({ metaRealmPath: META_REALM_PATH1, loadableRealmPath: LOADABLE_REALM_PATH1 })).rejects.toThrowError();
            await expect(LoadableRealmManager.reloadRealm({ metaRealmPath: META_REALM_PATH1, loadableRealmPath: LOADABLE_REALM_PATH1 })).rejects.toThrowError();

            expect(() => LoadableRealmManager.closeLoadableRealm({ metaRealmPath: META_REALM_PATH1, loadableRealmPath: LOADABLE_REALM_PATH1 })).not.toThrowError();
            expect(() => LoadableRealmManager.closeLoadableRealm({ metaRealmPath: META_REALM_PATH1, loadableRealmPath: LOADABLE_REALM_PATH1 })).not.toThrowError();
    });

    it('Should have no open LoadableRealms', async () => {

        expect(LoadableRealmManager.getAllLoadableRealmKeys()).toEqual([]);
        expect(LoadableRealmManager.getAllLoadableRealms()).toEqual([]);
        expect(LoadableRealmManager.getAllOpenLoadableRealmKeys()).toEqual([]);
        expect(LoadableRealmManager.getAllOpenLoadableRealmKeys()).toEqual([]);
    });

    it('Should be able to open 1 LoadableRealm with LoadableSchemas 1-3 in MetaRealm1.LoadableRealm1', async () => {

        await saveSchema({ metaRealmPath: META_REALM_PATH1, loadableRealmPath: LOADABLE_REALM_PATH1, schema: LOADABLE_SCHEMA_1 });
        await saveSchema({ metaRealmPath: META_REALM_PATH1, loadableRealmPath: LOADABLE_REALM_PATH1, schema: LOADABLE_SCHEMA_2 });
        await saveSchema({ metaRealmPath: META_REALM_PATH1, loadableRealmPath: LOADABLE_REALM_PATH1, schema: LOADABLE_SCHEMA_3 });

        const realm = await LoadableRealmManager.loadRealm({ metaRealmPath: META_REALM_PATH1, loadableRealmPath: LOADABLE_REALM_PATH1 });

        expect(() => LoadableRealmManager.getAllLoadableRealmKeys()).not.toThrowError();
        expect(LoadableRealmManager.getAllLoadableRealmKeys()).toEqual([ `${META_REALM_PATH1}-${LOADABLE_REALM_PATH1}` ]);
        expect(() => LoadableRealmManager.getAllLoadableRealms()).not.toThrowError();
        // TODO CHECK THIS; ADD SNAPSHOT
        expect(LoadableRealmManager.getAllLoadableRealms()).toMatchSnapshot();
        
        expect(() => LoadableRealmManager.getAllOpenLoadableRealmKeys()).not.toThrowError();
        expect(LoadableRealmManager.getAllOpenLoadableRealmKeys()).toEqual([ `${META_REALM_PATH1}-${LOADABLE_REALM_PATH1}` ]);
        expect(() => LoadableRealmManager.getAllOpenLoadableRealms()).not.toThrowError();
        // TODO CHECK THIS; ADD SNAPSHOT
        expect(LoadableRealmManager.getAllOpenLoadableRealms()).toMatchSnapshot();

        // TODO CHECK THIS; ADD SNAPSHOT
        expect(realm).toMatchSnapshot();

        const TEST_ROW1 = { name: 'Test1', schema1Data:'Hello!' };
        const TEST_ROW2 = { name: 'Test2', schema1Data:'World!' };

        realm.write(() => {
            realm.create(LOADABLE_SCHEMA_1.name, TEST_ROW1);
            realm.create(LOADABLE_SCHEMA_1.name, TEST_ROW2);
        });
        
        const results = realm.objects(LOADABLE_SCHEMA_1.name);

        expect(realm.objects(LOADABLE_SCHEMA_1.name).toJSON()).toEqual([ TEST_ROW1, TEST_ROW2 ]);
    });

    it('Should be able to open another LoadableRealm with LoadableSchemas 2-3 in MetaRealm2.LoadableRealm1', async () => {

        await saveSchema({ metaRealmPath: META_REALM_PATH2, loadableRealmPath: LOADABLE_REALM_PATH1, schema: LOADABLE_SCHEMA_2 });
        await saveSchema({ metaRealmPath: META_REALM_PATH2, loadableRealmPath: LOADABLE_REALM_PATH1, schema: LOADABLE_SCHEMA_3 });

        const realm = await LoadableRealmManager.loadRealm({ metaRealmPath: META_REALM_PATH2, loadableRealmPath: LOADABLE_REALM_PATH1 });

        expect(() => LoadableRealmManager.getAllLoadableRealmKeys()).not.toThrowError();
        expect(LoadableRealmManager.getAllLoadableRealmKeys()).toEqual([ `${META_REALM_PATH1}-${LOADABLE_REALM_PATH1}`, `${META_REALM_PATH2}-${LOADABLE_REALM_PATH1}` ]);
        expect(() => LoadableRealmManager.getAllLoadableRealms()).not.toThrowError();
        expect(LoadableRealmManager.getAllLoadableRealms()).toMatchSnapshot();
        
        expect(() => LoadableRealmManager.getAllOpenLoadableRealmKeys()).not.toThrowError();
        expect(LoadableRealmManager.getAllOpenLoadableRealmKeys()).toEqual([ `${META_REALM_PATH1}-${LOADABLE_REALM_PATH1}`, `${META_REALM_PATH2}-${LOADABLE_REALM_PATH1}` ]);
        expect(() => LoadableRealmManager.getAllOpenLoadableRealms()).not.toThrowError();
        expect(LoadableRealmManager.getAllOpenLoadableRealms()).toMatchSnapshot();

        expect(realm).toMatchSnapshot();

        const TEST_ROW21 = { name: 'Test21', schema2Data:'Hello!' };
        const TEST_ROW22 = { name: 'Test22', schema2Data:'World!' };
        const TEST_ROW31 = { name: 'Test31', schema3Data:'Hello!' };
        const TEST_ROW32 = { name: 'Test32', schema3Data:'World!' };

        realm.write(() => {
            realm.create(LOADABLE_SCHEMA_2.name, TEST_ROW21);
            realm.create(LOADABLE_SCHEMA_2.name, TEST_ROW22);

            realm.create(LOADABLE_SCHEMA_3.name, TEST_ROW31);
            realm.create(LOADABLE_SCHEMA_3.name, TEST_ROW32);
        });

        expect(realm.objects(LOADABLE_SCHEMA_2.name).toJSON()).toEqual([ TEST_ROW21, TEST_ROW22 ]);
        expect(realm.objects(LOADABLE_SCHEMA_3.name).toJSON()).toEqual([ TEST_ROW31, TEST_ROW32 ]);
    });

    it('Should be able to open another LoadableRealm with LoadableSchemas 4-5 in MetaRealm2.LoadableRealm2', async () => {

        await saveSchema({ metaRealmPath: META_REALM_PATH2, loadableRealmPath: LOADABLE_REALM_PATH2, schema: LOADABLE_SCHEMA_4 });
        await saveSchema({ metaRealmPath: META_REALM_PATH2, loadableRealmPath: LOADABLE_REALM_PATH2, schema: LOADABLE_SCHEMA_5 });

        const realm21 = await LoadableRealmManager.reloadRealm({ metaRealmPath: META_REALM_PATH2, loadableRealmPath: LOADABLE_REALM_PATH1 });

        const realm22 = await LoadableRealmManager.reloadRealm({ metaRealmPath: META_REALM_PATH2, loadableRealmPath: LOADABLE_REALM_PATH2 });

        expect(() => LoadableRealmManager.getAllLoadableRealmKeys()).not.toThrowError();
        expect(LoadableRealmManager.getAllLoadableRealmKeys()).toEqual([ `${META_REALM_PATH1}-${LOADABLE_REALM_PATH1}`, `${META_REALM_PATH2}-${LOADABLE_REALM_PATH1}`, `${META_REALM_PATH2}-${LOADABLE_REALM_PATH2}` ]);
        expect(() => LoadableRealmManager.getAllLoadableRealms()).not.toThrowError();
        expect(LoadableRealmManager.getAllLoadableRealms()).toMatchSnapshot();
        
        expect(() => LoadableRealmManager.getAllOpenLoadableRealmKeys()).not.toThrowError();
        expect(LoadableRealmManager.getAllOpenLoadableRealmKeys()).toEqual([ `${META_REALM_PATH1}-${LOADABLE_REALM_PATH1}`, `${META_REALM_PATH2}-${LOADABLE_REALM_PATH1}`, `${META_REALM_PATH2}-${LOADABLE_REALM_PATH2}` ]);
        expect(() => LoadableRealmManager.getAllOpenLoadableRealms()).not.toThrowError();
        expect(LoadableRealmManager.getAllOpenLoadableRealms()).toMatchSnapshot();

        expect(realm21).toMatchSnapshot();
        expect(realm22).toMatchSnapshot();

        const TEST_ROW21 = { name: 'Test21', schema2Data:'Hello!' };
        const TEST_ROW22 = { name: 'Test22', schema2Data:'World!' };
        const TEST_ROW31 = { name: 'Test31', schema3Data:'Hello!' };
        const TEST_ROW32 = { name: 'Test32', schema3Data:'World!' };
        const TEST_ROW41 = { name: 'Test41', schema4Data:'Hello!' };
        const TEST_ROW42 = { name: 'Test42', schema4Data:'World!' };
        const TEST_ROW51 = { name: 'Test51', schema5Data:'Hello!' };
        const TEST_ROW52 = { name: 'Test52', schema5Data:'World!' };


        realm21.write(() => {
            expect(() => realm21.create(LOADABLE_SCHEMA_2.name, TEST_ROW21)).toThrowError();
            expect(() => realm21.create(LOADABLE_SCHEMA_2.name, TEST_ROW22)).toThrowError();

            expect(() => realm21.create(LOADABLE_SCHEMA_3.name, TEST_ROW31)).toThrowError();
            expect(() => realm21.create(LOADABLE_SCHEMA_3.name, TEST_ROW32)).toThrowError();
        });

        realm22.write(() => {
            realm22.create(LOADABLE_SCHEMA_4.name, TEST_ROW41);
            realm22.create(LOADABLE_SCHEMA_4.name, TEST_ROW42);

            realm22.create(LOADABLE_SCHEMA_5.name, TEST_ROW51);
            realm22.create(LOADABLE_SCHEMA_5.name, TEST_ROW52);
        });

        expect(realm21.objects(LOADABLE_SCHEMA_2.name).toJSON()).toEqual([ TEST_ROW21, TEST_ROW22 ]);
        expect(realm21.objects(LOADABLE_SCHEMA_3.name).toJSON()).toEqual([ TEST_ROW31, TEST_ROW32 ]);
        expect(realm22.objects(LOADABLE_SCHEMA_4.name).toJSON()).toEqual([ TEST_ROW41, TEST_ROW42 ]);
        expect(realm22.objects(LOADABLE_SCHEMA_5.name).toJSON()).toEqual([ TEST_ROW51, TEST_ROW52 ]);
    });

    afterAll(async () => {
        await LoadableRealmManager.closeAll();
        await MetaRealmManager.closeAll();
    });
});
