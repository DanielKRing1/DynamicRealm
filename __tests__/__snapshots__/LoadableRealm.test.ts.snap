// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoadableRealmManager Should be able to open 1 LoadableRealm with LoadableSchemas 1-3 in MetaRealm1.LoadableRealm1 1`] = `
Array [
  Realm {},
]
`;

exports[`LoadableRealmManager Should be able to open 1 LoadableRealm with LoadableSchemas 1-3 in MetaRealm1.LoadableRealm1 2`] = `
Array [
  Realm {},
]
`;

exports[`LoadableRealmManager Should be able to open 1 LoadableRealm with LoadableSchemas 1-3 in MetaRealm1.LoadableRealm1 3`] = `Realm {}`;

exports[`LoadableRealmManager Should be able to open another LoadableRealm with LoadableSchemas 2-3 in MetaRealm2.LoadableRealm1 1`] = `
Array [
  Realm {},
  Realm {},
]
`;

exports[`LoadableRealmManager Should be able to open another LoadableRealm with LoadableSchemas 2-3 in MetaRealm2.LoadableRealm1 2`] = `
Array [
  Realm {},
  Realm {},
]
`;

exports[`LoadableRealmManager Should be able to open another LoadableRealm with LoadableSchemas 2-3 in MetaRealm2.LoadableRealm1 3`] = `Realm {}`;

exports[`LoadableRealmManager Should be able to open another LoadableRealm with LoadableSchemas 4-5 in MetaRealm2.LoadableRealm2 1`] = `
Array [
  Realm {},
  Realm {},
  Realm {},
]
`;

exports[`LoadableRealmManager Should be able to open another LoadableRealm with LoadableSchemas 4-5 in MetaRealm2.LoadableRealm2 2`] = `
Array [
  Realm {},
  Realm {},
  Realm {},
]
`;

exports[`LoadableRealmManager Should be able to open another LoadableRealm with LoadableSchemas 4-5 in MetaRealm2.LoadableRealm2 3`] = `Realm {}`;

exports[`LoadableRealmManager Should be able to open another LoadableRealm with LoadableSchemas 4-5 in MetaRealm2.LoadableRealm2 4`] = `Realm {}`;
